craft_application/services/linter.py

```diff
@@ -0,0 +1,268 @@
+            if linter_name not in base_cfg:
+                base_cfg[linter_name] = IgnoreSpec(
+                    ids=set() if over_spec.ids != "*" else "*",
+                    by_filename={k: set(v) for k, v in over_spec.by_filename.items()},
+                )
+                if over_spec.ids == "*":
+                    continue
+                if isinstance(over_spec.ids, set):
+                    base_cfg[linter_name].ids = set(over_spec.ids)
+                continue
+            base_spec = base_cfg[linter_name]
+
+            if over_spec.ids == "*":
+                base_spec.ids = "*"
+                base_spec.by_filename.clear()
+            elif isinstance(over_spec.ids, set):
+                if base_spec.ids != "*":
+                    if not isinstance(base_spec.ids, set):
+                        base_spec.ids = set()
+                    base_spec.ids.update(over_spec.ids)
+            for issue_id, globs in over_spec.by_filename.items():
+                base_spec.by_filename.setdefault(issue_id, set()).update(globs)
+
+    def pre_filter_linters(
+        self,
+        stage: Stage,
+        ctx: LintContext,  # noqa: ARG002 (reserved for future use)
+        candidates: list[type[AbstractLinter]] | None = None,
+    ) -> list[type[AbstractLinter]]:
+        """App-specific selection hook."""
+        registry = type(self)._class_registry  # noqa: SLF001
+        return list(candidates or registry.get(stage, []))
+
+    def post_filter_issues(
+        self,
+        linter: AbstractLinter,  # noqa: ARG002 (reserved for future use)
+        issues: Iterable[LinterIssue],
+        ctx: LintContext,  # noqa: ARG002 (reserved for future use)
+    ) -> Iterator[LinterIssue]:
+        """App-specific filtering hook."""
+        yield from issues
+
+    def run(
+        self,
+        stage: Stage,
+        ctx: LintContext,
+    ) -> Iterator[LinterIssue]:
+        """Run linters for a stage, streaming non-suppressed issues."""
+        self._issues.clear()
+        registry = type(self)._class_registry  # noqa: SLF001
+        selected = self.pre_filter_linters(stage, ctx, registry.get(stage, []))
+        for cls in selected:
+            linter = cls()
+            raw_issues = linter.run(ctx)
+
+            user_filtered = (
+                issue
+                for issue in raw_issues
+                if not should_ignore(linter.name, issue, self._ignore_cfg)
+            )
+            filtered = self.post_filter_issues(linter, user_filtered, ctx)
+            for issue in filtered:
+                self._issues.append(issue)
+                yield issue
+
+    def summary(self) -> ExitCode:
+        """Summarize results as an ExitCode based on highest severity."""
+        if any(i.severity == Severity.ERROR for i in self._issues):
+            return ExitCode.ERROR
+        if any(i.severity == Severity.WARNING for i in self._issues):
+            return ExitCode.WARN
+        return ExitCode.OK
```

Comment:
This would probably be better as a `get_highest_severity` or similar. If we run linters during a lifecycle command (as we plan to), we don't want to exit nonzero on warnings.